File: packages/client/src/services/recordService.ts
import { BaseService } from './baseService.js';
import type { BaseForge } from '../client.js';
import type {
  RecordModel,
  ListResult,
  ListOptions,
  GetOneOptions,
} from '../types.js';
import { ClientResponseError } from '../types.js';

export class RecordService extends BaseService {
  readonly collectionName: string;

  constructor(client: BaseForge, collectionName: string) {
    super(client);
    this.collectionName = collectionName;
  }

  /** Path dasar endpoint koleksi */
  protected get basePath(): string {
    return `api/p/${this.client.projectId}/collections/${this.collectionName}/records`;
  }

  /**
   * Mengambil daftar record dengan pagination, sorting, filtering, & expand
   */
  async getList<T = RecordModel>(
    page = 1,
    perPage = 20,
    options: ListOptions = {}
  ): Promise<ListResult<T>> {
    const params: Record<string, unknown> = {
      page,
      perPage,
      sort: options.sort,
      filter: options.filter,
      search: options.search,
      expand: options.expand,
    };

    return this.request<ListResult<T>>(this.basePath, {
      method: 'GET',
      params,
      headers: options.headers,
    });
  }

  /**
   * Mengambil SELURUH record dalam koleksi (otomatis mem-page hingga selesai)
   */
  async getFullList<T = RecordModel>(
    batchSize = 200,
    options: Omit<ListOptions, 'page' | 'perPage'> = {}
  ): Promise<T[]> {
    let page = 1;
    const allItems: T[] = [];

    while (true) {
      const res = await this.getList<T>(page, batchSize, options);
      allItems.push(...res.items);
      if (res.page >= res.totalPages || res.items.length === 0) {
        break;
      }
      page++;
    }

    return allItems;
  }

  /**
   * Mengambil record pertama yang cocok dengan filter yang diberikan
   */
  async getFirstListItem<T = RecordModel>(
    filter: string,
    options: Omit<ListOptions, 'filter' | 'page' | 'perPage'> = {}
  ): Promise<T> {
    const res = await this.getList<T>(1, 1, { ...options, filter });
    if (!res.items || res.items.length === 0) {
      throw new ClientResponseError(404, `No record found matching filter: "${filter}"`, null, 'NOT_FOUND');
    }
    return res.items[0];
  }

  /**
   * Mengambil satu baris record berdasarkan ID
   */
  async getOne<T = RecordModel>(id: string, options: GetOneOptions = {}): Promise<T> {
    const res = await this.request<{ record: T }>(`${this.basePath}/${encodeURIComponent(id)}`, {
      method: 'GET',
      params: { expand: options.expand },
      headers: options.headers,
    });
    return res.record;
  }

  /**
   * Membuat record baru (menerima JSON object atau FormData untuk upload file).
   * M34: kalau data punya field `id`, server akan pakai itu sebagai document ID
   * (untuk migration dari Appwrite/PocketBase yang punya existing IDs).
   */
  async create<T = RecordModel>(data: Record<string, unknown> | FormData): Promise<T> {
    const res = await this.request<{ record: T }>(this.basePath, {
      method: 'POST',
      body: data as BodyInit,
    });
    return res.record;
  }

  /**
   * M34: Membuat record dengan custom document ID.
   * ID harus 1-64 karakter [a-zA-Z0-9_-]. Duplikat → 409 DOCUMENT_ID_TAKEN.
   */
  async createWithId<T = RecordModel>(
    id: string,
    data: Record<string, unknown>
  ): Promise<T> {
    return this.create<T>({ ...data, id });
  }

  /**
   * Mengubah record yang sudah ada
   */
  async update<T = RecordModel>(id: string, data: Record<string, unknown> | FormData): Promise<T> {
    const res = await this.request<{ record: T }>(`${this.basePath}/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      body: data as BodyInit,
    });
    return res.record;
  }

  /**
   * Menghapus record berdasarkan ID
   */
  async delete(id: string): Promise<boolean> {
    await this.request(`${this.basePath}/${encodeURIComponent(id)}`, {
      method: 'DELETE',
    });
    return true;
  }

  /**
   * Autentikasi ke Auth Collection menggunakan identity/email & password (PocketBase Parity)
   */
  async authWithPassword<T = RecordModel>(
    identity: string,
    password: string
  ): Promise<{ token: string; refreshToken?: string; record: T }> {
    const res = await this.request<{ token: string; refreshToken?: string; record: T }>(
      `api/p/${this.client.projectId}/collections/${this.collectionName}/auth-with-password`,
      {
        method: 'POST',
        body: { identity, password },
      }
    );
    this.client.authStore.save(res.token, res.refreshToken || '', res.record as unknown as import('../types.js').AuthUser);
    return res;
  }

  /**
   * Memperbarui token sesi pengguna saat ini
   */
  async authRefresh<T = RecordModel>(): Promise<{ token: string; refreshToken?: string; record: T }> {
    const res = await this.request<{ token: string; refreshToken?: string; record: T }>(
      `api/p/${this.client.projectId}/collections/${this.collectionName}/auth-refresh`,
      {
        method: 'POST',
      }
    );
    this.client.authStore.save(res.token, res.refreshToken || '', res.record as unknown as import('../types.js').AuthUser);
    return res;
  }
}

File: packages/kotlin-client/src/main/kotlin/id/wekanz/baseforge/records/RecordService.kt
package id.wekanz.baseforge.records

import id.wekanz.baseforge.ClientResponseException
import id.wekanz.baseforge.internal.BaseForgeJson
import id.wekanz.baseforge.internal.HttpCore
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * Operasi CRUD untuk satu koleksi — paritas dengan `recordService.ts` di SDK TS.
 *
 * Didapat lewat [id.wekanz.baseforge.BaseForge.collection], jangan di-new langsung.
 */
public class RecordService internal constructor(
    private val http: HttpCore,
    private val projectId: String,
    public val collectionName: String,
) {
    private val basePath: String get() = "api/p/$projectId/collections/$collectionName/records"

    /** Ambil daftar record dengan pagination/sort/filter/search/expand. */
    public suspend fun getList(
        page: Int = 1,
        perPage: Int = 20,
        options: ListOptions = ListOptions(),
    ): RecordList {
        val query = mapOf(
            "page" to page.toString(),
            "perPage" to perPage.toString(),
            "sort" to options.sort,
            "filter" to options.filter,
            "search" to options.search,
            "expand" to options.expand,
        )
        val text = http.request("GET", basePath, query = query) ?: "{}"
        val raw = BaseForgeJson.decodeFromString(ListResultRaw.serializer(), text)
        return RecordList(
            page = raw.page,
            perPage = raw.perPage,
            totalItems = raw.totalItems,
            totalPages = raw.totalPages,
            items = raw.items.map { jsonToRecord(it) },
        )
    }

    /**
     * Ambil record pertama yang cocok dengan [filter].
     * Melempar `404 NOT_FOUND` bila tidak ada — paritas `getFirstListItem` di SDK TS.
     */
    public suspend fun getFirstListItem(filter: String, expand: String? = null): RecordModel {
        val res = getList(1, 1, ListOptions(filter = filter, expand = expand))
        return res.items.firstOrNull()
            ?: throw ClientResponseException(404, "No record found matching filter: \"$filter\"", "NOT_FOUND")
    }

    /** Ambil satu record by id. 404 bila tidak ada. */
    public suspend fun getOne(id: String, expand: String? = null): RecordModel {
        val text = http.request("GET", "$basePath/$id", query = mapOf("expand" to expand))
            ?: error("empty getOne response")
        val env = BaseForgeJson.decodeFromString(RecordEnvelope.serializer(), text)
        return jsonToRecord(env.record)
    }

    /**
     * Buat record baru. Bila [data] punya `id`, server memakainya sebagai document ID
     * (untuk migrasi dengan existing IDs; duplikat → 409 DOCUMENT_ID_TAKEN).
     */
    public suspend fun create(data: JsonObject): RecordModel {
        val text = http.request("POST", basePath, body = data.toString()) ?: error("empty create response")
        val env = BaseForgeJson.decodeFromString(RecordEnvelope.serializer(), text)
        return jsonToRecord(env.record)
    }

    /** Buat record dengan custom document ID (M34). */
    public suspend fun createWithId(id: String, data: JsonObject): RecordModel {
        val merged = buildJsonObject {
            data.forEach { (k, v) -> put(k, v) }
            put("id", JsonPrimitive(id))
        }
        return create(merged)
    }

    /** Ubah record yang sudah ada. */
    public suspend fun update(id: String, data: JsonObject): RecordModel {
        val text = http.request("PATCH", "$basePath/$id", body = data.toString())
            ?: error("empty update response")
        val env = BaseForgeJson.decodeFromString(RecordEnvelope.serializer(), text)
        return jsonToRecord(env.record)
    }

    /** Hapus record. */
    public suspend fun delete(id: String): Boolean {
        http.request("DELETE", "$basePath/$id")
        return true
    }
}
